### ISSUE TITLE
Alternate DatasetLoader For Training Custom Data

### ISSUE THREAD

**Issue by afiaka87** [lucidrains/DALLE-pytorch#68]
I saw this in the README.md.
> If you have a dataset with its own directory structure for tying together image and text descriptions, do let me know in the issues, and I'll see if I can accommodate it in the script.

This DataSet takes a folder for images and a dictionary mapping those images' filenames to the text they should be trained on. Preparation for loading would basically be:

- get images
- get labels
- create a csv, or in memory python object mapping image filenames to labels
- pass it in to `TextImagePairs`

e.g.:
|`image_filename`        |    `text`        |
|----------------|-----------------|
|"abc.png"    |   "an image of a squirrel"|
|"xyz.png"    |   "an image of a peanut"|
|"etc.png"     |   "etc..."        |

In this example I have a pre-prepared list of image-to-text pairs in a parquet file. 
All I have to do to pass it in is import pandas, load it, convert it to a dictionary and then pass it in TextImagePairs
```python
import pandas as pd
from collections import OrderedDict
IMAGE_TEXT_FOLDER = './images'
current_image_subset = all_img_filenames
file_text_pairs_dict = pd.read_parquet("file_text_pairs.parq").to_dict(into=OrderedDict)
```

```python
'''
folder: str or Path to a flat directory containing all of your images.
image_text_pairs: a dict in the form 
{
    "filename.png": "a text description about the image at filename" 
    "cat.png": "an image of a cat staring",
    "etc.png": "...",
}
'''
class TextImagePairs(Dataset):
    def __init__(self, image_text_pairs, folder="./images", text_len = 256, image_size = 256):
        super().__init__()
        self._path = Path(folder)
        self.image_text_pairs = image_text_pairs
        self.keys = list(self.image_text_pairs)
        self.image_tranform = ImageTransform = T.Compose([
            T.CenterCrop(image_size),
            T.Resize(image_size),
            T.ToTensor(),
            T.Lambda(lambda t: t.expand(3, -1, -1)),
            T.Normalize((0.5,) * 3, (0.5,) * 3)
        ])
            
    def __getitem__(self, ind):
        image_filename = self.keys[ind]
        description = self.image_text_pairs[image_filename]
        image = Image.open(os.path.join(self._path, image_filename))
        tokenized_text = tokenize(description).squeeze(0)
        mask = tokenized_text != 0
        image_tensor = self.image_tranform(image)
        return tokenized_text, image_tensor, mask

    def __len__(self):
        return len(self.keys)
```

**Comment by afiaka87** [comment]
This method is annoyingly slow unfortunately. There's a hard CPU bottleneck whenever the function `text_from_filename` is called. I've indexed on the filename and added python's lru_cache with an unbound size, but still can't get it performing quickly. If anyone is more familiar with pandas and knows how I could speed this up, I'd appreciate it.

**Comment by sorrge** [comment]
Pandas is very slow extracting data from a dataframe like that. Dump it in a raw python dict and it will be x10000 times faster.

**Comment by afiaka87** [comment]
Thanks! This solved my bottleneck completely. Don't know why I didn't realize I could just call .to_dict().  I'll update my code in the original answer.

**Comment by afiaka87** [comment]
I think this could be useful for some. I definitely don't find the default image loader useful for anything other than COCO style datasets. I suppose the idea is that you need variety in your phrasing, but I don't think that's really true with pretrained CLIP in the mix.
